---
title: "Population genomics analysis"
author: "Imogen Foote (with scripts from Tom Oosting)"
date: "2023-05-24"
output: html_document
---

Load libraries
```{r}
rm(list=ls())
library(SNPRelate)
library(gdsfmt)
library(glue)
library(tidyverse)
library(poppr)
library(scales)
library(ggrepel)
library(pheatmap)
library(adegenet)
library(ggsignif)
library(dartR)


source("C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Thesis/Chapter 3_comparative genomics/Methods/Analysis/scripts/R_functions/SNPrelate_functions.R")
```

Set variables
```{r}
Project   <- "AllAlbatross"
data_set  <- "full"
filter    <- "neutral"

POP1 <- "ANT"
POP2 <- "GIB+DI"

snpfile_ext <- glue("C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Thesis/Chapter 3_comparative genomics/Methods/Analysis/data/{Project}/{Project}_{data_set}_{filter}")
outfile_ext <- glue("C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Thesis/Chapter 3_comparative genomics/Methods/Analysis/output/{Project}/{filter}_analyses/{Project}_{data_set}_{filter}/{POP1}_{POP2}/{POP1}_{POP2}")
dir.create(dirname(outfile_ext), recursive = TRUE)

```
Load in data
```{r}
#Load GDS file
gds <- snpgdsReadGDS(vcf_file = paste0(snpfile_ext,"_",POP1,"_",POP2,".vcf.gz"), 
                     gds_file = paste0(snpfile_ext,"_",POP1,"_",POP2,".gds"))


#Load sample information
#A tsv file with columns for sample names (IND) & population name (POP)
sample_info <- read_tsv(glue::glue("C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Thesis/Chapter 3_comparative genomics/Methods/Analysis/resources/{Project}_site_info.tsv"))
#add spp column
sample_info$SPP <- gsub("_.*", "", sample_info$IND)
#add abbreviated pop column for plotting
sample_info <- sample_info %>% 
  mutate(pop2 = case_when(
  POP == "Dougal" ~ "DG",
  POP == "Mt Waterhouse" ~ "MW",
  POP == "North Plains" ~ "NP",
  POP == "Orde Lees Stream" ~ "OR", 
  POP == "Study Area" ~ "SA",
  POP == "Amherst-Astrolabe" ~ "AA",
  POP == "Disappointment Island" ~ "DI",
  POP == "Mount Dick" ~ "MD",
  POP == "Maclaren Ridge" ~ "MR")) %>% 
  mutate(island = case_when(
  POP == "Dougal" ~ "Antipodes",
  POP == "Mt Waterhouse" ~ "Antipodes",
  POP == "North Plains" ~ "Antipodes",
  POP == "Orde Lees Stream" ~ "Antipodes", 
  POP == "Study Area" ~ "Antipodes",
  POP == "Amherst-Astrolabe" ~ "Adams",
  POP == "Disappointment Island" ~ "Disappointment",
  POP == "Mount Dick" ~ "Adams",
  POP == "Maclaren Ridge" ~ "Adams"  
  ))


###JUST FOR GIB-DI comparison####
# sample_info <- sample_info %>%
#   mutate(SPP = if_else(pop2 == "DI", "DI", SPP))

#subset sample info to keep only rows where 'IND' is also found in sample.id of gds file (shouldn't change anything because should all be in there but good check)
sample_info <- sample_info %>% filter(IND %in% read.gdsn(index.gdsn(gds, "sample.id")))

#read genlight file
if(file.exists(paste0(snpfile_ext,"_",POP1,"_",POP2,".genlight.R"))){
  load(paste0(snpfile_ext,"_",POP1,"_",POP2,".genlight.R"))
  strata(genlight) <- sample_info
} else {
  vcfR     <- vcfR::read.vcfR(paste0(snpfile_ext,"_",POP1,"_",POP2,".vcf.gz"))
  genlight <- vcfR::vcfR2genlight(vcfR)
  strata(genlight) <- sample_info
  save(genlight, file = paste0(snpfile_ext,"_",POP1,"_",POP2,".genlight.R"))
}

#check files
read.gdsn(index.gdsn(gds, "sample.id"))
length(read.gdsn(index.gdsn(gds, "sample.id")))

genlight@ind.names
length(genlight@ind.names)
```


Set up colours and shapes for plotting
```{r}
###BY POP

hex <- c(
  "Dougal" = "#6536bcff", 
  "Mt Waterhouse" = "#0878b9ff", 
  "North Plains" = "#151eadff",
  "Orde Lees Stream" = "#5e9a3fff",
  "Study Area" = "#49bc94ff",
    "Amherst-Astrolabe" = "#d82a43ff", 
  "Disappointment Island" = "#ee919dff",
  "Mount Dick" = "#f6991aff",
  "Maclaren Ridge" = "#f48156ff"
)


Npops      <- length(unique(sample_info$POP))
shapes     <- c(rep(c(15,16,17,18),10))[1:Npops]

# Ensure the sample_info POP column is a factor with levels in the desired order
sample_info <- sample_info %>%
  mutate(POP = factor(POP, levels = unique(POP)))

# Map the specific colors to the populations
pop_cols <- sample_info %>%
  group_by(POP) %>%
  summarise(N = n()) %>%
  mutate(color = hex[as.character(POP)],
         shape = shapes)



# #make set color df
# Npops      <- length(unique(sample_info$POP))
# hex_codes2 <- hue_pal()(Npops)
# shapes     <- c(rep(c(15,16,17,18),10))[1:Npops]
# 
# #set colors and node shpes for pops
# pop_cols <- sample_info   %>% 
#             mutate(POP = factor(POP, levels = unique(POP))) %>% ##this ensures it retains the order of pops from sample_info rather than alphabetically
#             group_by(POP) %>% 
#             summarise(N = n()) %>% 
#             mutate(color = hex_codes2,
#                    shape = shapes)

#filter rows of pop_cols to keep only those where 'POP' column values are found in sample_info$POP (again should be all, but check)
pop_cols_select <- pop_cols %>% filter(POP %in% sample_info$POP)
colors <- pop_cols_select$color
shapes <- pop_cols_select$shape

###BY SPP
spp_colors <- c("#2c6b94ff", "#ee525fff")
spp_shapes <- c(16,17)

##By island
#island_colors <- c("#00C19F80", "#FF61C380", "#b31f6d80")
island_colors <- c("#2c6b94ff", "#ee525fff", "#e8c0b8ff")
island_shapes <- c(16,17, 15)

```


Get some summary stats from the GDS file
```{r}
snpgdsSummary(gds)

#calculate per SNP info (missing rate, MAF, heterozygosity, ref allele count, alt allele count, ref allele freq, alt allele freq, and missing rate again - not sure how different from first)
bim <- snpgdsSNPsum(gds = gds, extended = TRUE)

#calculate heterozygosity and missing data per individual
sample_info$H <- snpgdsINDHet(gds = gds)
sample_info$MISS_IND <- snpgdsINDMiss(gds = gds)
ggplot(sample_info, aes(x=IND,y=MISS_IND))+
  geom_point()


#get chromosome information (chr length, cumulative length, LG) for downstream analysis incl. manhattan plots
#requires columns CHR and POS
chr_info <- chr_info(bim)    

#add chr_info to SNP info (bim)
#cumulative base pair (BPcum) is calculated for manhattan plotting
bim <- left_join(bim,chr_info[,c("CHR","tot")]) %>% 
       arrange(LG,POS)                               %>% 
       mutate(BPcum = tot + POS)

#same as snpgdsSNPSUM function above but splits full dataset out to give information about SNPs per population
sum_POP <- snpgdsSNPsum_byPOP(gds = gds, pop.id = sample_info$POP)
sum_SPP <- snpgdsSNPsum_byPOP(gds = gds, pop.id = sample_info$SPP)

  ggplot(sum_SPP, aes(x=POP, y=HET, fill=POP)) +
    geom_boxplot(show.legend=FALSE) +
    scale_fill_manual(values=c("darkcyan", "palevioletred")) +
    labs(x = "", y = "Heterozygosity per SNP") +
    theme_bw() +
    theme(axis.text = element_text(size = 14),
          axis.title = element_text(size = 14))
  
    ggplot(sum_SPP, aes(x=POP, y=HET, fill=POP)) +
    geom_boxplot(show.legend=FALSE) +
    scale_fill_manual(values=spp_colors) +
    labs(x = "", y = "Heterozygosity per SNP") +
    theme_bw() +
    theme(axis.text = element_text(size = 14),
          axis.title = element_text(size = 14))

##NOT QUITE SURE WHAT THIS WILL BE USED FOR ####
#calculate minor allele freq (maf) or minor allele count (mac) for each snp per population, and number of populations with maf/mac (final column)
#this assumes that the alt allele is always the minor allele 
maf_POP <- snpgdsSNPmaf_byPOP(gds = gds, pop.id = sample_info$POP, mac = TRUE)
#create new df with number of pops (N) that share a given maf/mac, then gives the count of maf/mac shared 
shared <- maf_POP %>% group_by(N) %>% summarise(n = n()) %>% arrange(desc(N))





```

```{r}
pop(genlight) <- sample_info$SPP
genlight <- gl.compliance.check(genlight)
Het <- gl.report.heterozygosity(genlight, method = "pop")
write.table(Het, file=glue("{outfile_ext}_heterozygosity.txt"))
```



Look at heterozygosity and perform ANOVA to look at significant differences
```{r}
###BY SPP###
#print heterozygosity
spp.het <- sample_info %>% 
  group_by(SPP) %>% 
  summarise(mean_H = mean(H),
            sd_H = sd(H),
            min_H = min(H),
            max_H = max(H))

write.table(spp.het, file=glue("{outfile_ext}_heterozygosity_by_SPP.txt"))

#perform ANOVA to look at differences in het between different pops
spp.res.aov <- aov(H ~ SPP, data = sample_info)

#check normality
plot(spp.res.aov, 2)

#see if significant
summary(spp.res.aov) 

# if so, which comparisons are significant
spp.tukey <- TukeyHSD(spp.res.aov) 

#create list of significant comparisons
spp.tukey_list <- str_split(rownames(spp.tukey$SPP)[which(spp.tukey$SPP[,4] < 0.05)],"-") 


#calculate pop means for plotting
spp.mean <- sample_info %>% 
  group_by(SPP) %>% 
  summarise(mean_H = mean(H))

#plot boxplot showing significant comparisons 
if (length(spp.tukey_list) > 0) {
  #create boxplot with significant comparisons
  ggplot(sample_info, aes(x=SPP, y=H, fill=SPP))+
    geom_boxplot(show.legend=FALSE)+
    geom_signif(comparisons = spp.tukey_list,
                map_signif_level = TRUE,
                test = "wilcox.test",
                step_increase = 0.07)+
    scale_fill_manual(values=spp_colors) +
    labs(x = "",
          y = "Heterozygosity") +
    theme_bw() +
    theme(axis.text = element_text(size = 14), 
          axis.title=element_text(size=14))
ggsave(filename = glue("{outfile_ext}_Het_boxplot_bySPP.png"), dpi=300)
} else {
  #create boxplot without significant comparisons
  ggplot(sample_info, aes(x=SPP, y=H, fill=SPP)) +
    geom_boxplot(show.legend=FALSE) +
    scale_fill_manual(values=spp_colors) +
    labs(x = "", y = "Heterozygosity") +
    theme_bw() +
    theme(axis.text = element_text(size = 14),
          axis.title = element_text(size = 14)) 
ggsave(filename = glue("{outfile_ext}_Het_boxplot_bySPP.png"), dpi=300)
}


# ggplot(sample_info, aes(x=POP,y=H))+
#   geom_boxplot()+
#   geom_signif(comparisons = tukey_list,
#               map_signif_level = FALSE,
#               test = "wilcox.test",
#               step_increase = 0.07)+
#   labs(x = "sample location",
#        y = "heterozygosity")+
#   theme_bw()
# ggsave(filename = glue("{outfile_ext}_Het_boxplot2.png"))


###BY POP### 
#print heterozygosity
pop.het <- sample_info %>% 
  group_by(POP) %>% 
  summarise(mean_H = mean(H))

write.table(pop.het, file=glue("{outfile_ext}_heterozygosity_by_POP.txt"))


#perform ANOVA to look at differences in get between different pops
pop.res.aov <- aov(H ~ POP, data = sample_info)

#check normality
plot(pop.res.aov, 2)

#see if significant
summary(pop.res.aov) 

# if so, which comparisons are significant
pop.tukey <- TukeyHSD(pop.res.aov) 

#create list of significant comparisons
pop.tukey_list <- str_split(rownames(pop.tukey$POP)[which(pop.tukey$POP[,4] < 0.05)],"-") 

#plot boxplot showing significant comparisons 
if (length(pop.tukey_list) > 0) {
  #create boxplot with significant comparisons
  ggplot(sample_info, aes(x=factor(pop2, levels = unique(pop2)), y=H, fill=pop2))+
    geom_boxplot(show.legend=FALSE)+
    geom_signif(comparisons = pop.tukey_list,
                map_signif_level = TRUE,
                test = "wilcox.test",
                step_increase = 0.07)+
    scale_fill_manual(values=colors) +
    labs(x = "",
          y = "Heterozygosity") +
    theme_bw() +
    theme(axis.text = element_text(size = 12), 
          axis.title=element_text(size=14))
ggsave(filename = glue("{outfile_ext}_Het_boxplot_byPOP.png"), dpi=300)
} else {
  #create boxplot without significant comparisons
  ggplot(sample_info, aes(x=factor(pop2, levels = unique(pop2)), y=H, fill=pop2)) +
    geom_boxplot(show.legend=FALSE) +
    scale_fill_manual(values=colors) +
    labs(x = "", y = "Heterozygosity") +
    theme_bw() +
    theme(axis.text.x = element_text(size = 12),
          axis.title = element_text(size = 14))
ggsave(filename = glue("{outfile_ext}_Het_boxplot_byPOP.png"), dpi=300)
}


```

AMOVA
```{r}
#check ploidy of genlight file and set to diploid if it has incorrectly assigned it https://groups.google.com/g/poppr/c/FNajXuqyWsc
ploidy(genlight)
ploidy(genlight) <- 2

#perform AMOVA POP level
AMOVA.POP <- poppr::poppr.amova(genlight, ~POP, missing = "ignore")
capture.output(AMOVA.POP, file = glue::glue("{outfile_ext}.AMOVA.POP.txt") )
#significance test
AMOVA.POP_significance <- ade4::randtest(AMOVA.POP, nrepet = 999)

AMOVA.POP_significance$names
AMOVA.POP_significance$pvalue

capture.output(AMOVA.POP_significance, file = glue::glue("{outfile_ext}.AMOVA.POP_significance.txt"))
save(AMOVA.POP,AMOVA.POP_significance,file = glue::glue("{outfile_ext}.AMOVA.POP.R"))
##################################################
#copy into console below to print full results#
#then copy into significance.txt output #
AMOVA.POP_significance                               #
#don't know why in prints different from script  #
##################################################
#print significance results
png(filename = glue::glue("{outfile_ext}.AMOVA.POP_significance_plot.png"), res = 300, height = 4, width = 8 , units = "in"  )
  plot(AMOVA.POP_significance)
dev.off()


#perform AMOVA SPP level
AMOVA.SPP <- poppr::poppr.amova(genlight, ~SPP)
capture.output(AMOVA.SPP, file = glue::glue("{outfile_ext}.AMOVA.SPP.txt"))
#significance test
AMOVA.SPP_significance <- ade4::randtest(AMOVA.SPP, nrepet = 999)

AMOVA.SPP_significance$names
AMOVA.SPP_significance$pvalue



capture.output(AMOVA.SPP_significance, file = glue::glue("{outfile_ext}.AMOVA.SPP_significance.txt"))
save(AMOVA.SPP,AMOVA.SPP_significance,file = glue::glue("{outfile_ext}.AMOVA.SPP.R"))
##################################################
#copy into console below to print full results   #
AMOVA.SPP_significance                               #
#don't know why in prints different from script  #
##################################################
#print significance results
png(filename = glue::glue("{outfile_ext}.AMOVA.SPP_significance_plot.png"), res = 300, height = 4, width = 8 , units = "in"  )
  plot(AMOVA.SPP_significance)
dev.off()
```

PCA
```{r}
###By POP
#estimate genetic relationship matrix
GRM <- snpgdsGRM(gds, sample.id = sample_info$IND, method = "Eigenstrat", autosome.only = FALSE)


#perform and plot PCA - note if you want to change the legend title you have to change it within the PCA_manual_colors function
png(filename =  glue::glue("{outfile_ext}_GRM.POP.png"), width = 7, height = 4, units = "in", res = 300) 
  PCA_SNA  <- PCA_manual_colors(GRM$grm, 
                                sample_info$IND,
                                sample_info$POP,
                                1, 2,            #PCs to plot
                                show.point=T, 
                                show.label=F, 
                                manual_colors = colors, 
                                manual_shapes = shapes,
                                show.ellipse=T, 
                                show.line=F, 
                                alpha=0) 
dev.off()



#perform classic PCA - results should the same as the GRM
pca_snprelate <- snpgdsPCA(gds, autosome.only = FALSE)
#extract variance proportions explained by each PC and convert to %
pc.percent <- pca_snprelate$varprop*100
head(round(pc.percent, 2))

#screeplot
variance_proportions <- pca_snprelate$varprop

# Create a scree plot
ggplot(data = data.frame(Principal_Component = 1:length(variance_proportions), Variance_Explained = variance_proportions), aes(x = Principal_Component, y = Variance_Explained)) +
  geom_point(stat = "identity", fill = "skyblue") +
  geom_point(color = "blue", size = 3) +
  labs(x = "Principal Component", y = "Variance Explained (%)", title = "Scree Plot") +
  theme_minimal()

#first make sure populations are in same order as data frame for plotting/legend order
sample_info$POP <- factor(sample_info$POP, levels = unique(sample_info$POP))
#create data frame containing sample.id, population and EV1 and EV2 (first and second eigenvectors from pca_snprelate)
tab <- data.frame(sample.id  = pca_snprelate$sample.id,
                  region     = sample_info$POP,
                  EV1        = pca_snprelate$eigenvect[,1],    # the first eigenvector
                  EV2        = pca_snprelate$eigenvect[,2],    # the second eigenvector
                  stringsAsFactors = FALSE)
#selects values where EV1 <= -0.1 and EV2 <= -0.3, but not sure why
tab_select <- dplyr::filter(tab,EV1 <=  -0.1| EV2 <= -0.3)

png(filename =  glue::glue("{outfile_ext}_PCA.POP.png"), width = 8, height = 4, units = "in", res = 300) 
  ggplot(tab,aes(x=EV1,y=EV2, color = region ))+
    geom_point()+
    geom_text_repel(data = tab_select,aes(x=EV1,y=EV2, color = tab_select$region, label = tab_select$sample.id), show.legend = F,max.overlaps = 30)+
    xlab(glue("PC1 ({round(pc.percent[1],2)}%)"))+
    ylab(glue("PC2 ({round(pc.percent[2],2)}%)")) +
    theme(
      panel.border = element_rect(color = "black", fill = NA, linewidth = 0.75),
      panel.background = element_blank(),
      panel.grid.major = element_line(color = "grey", linewidth = 0.25),
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      legend.key = element_blank()) +
    labs(color = "Sampling site")
dev.off()
  
#calculate SNP loadings from PCA
pca_snprelate_load  <- snpgdsPCASNPLoading(pca_snprelate, gds, num.thread=1L, verbose=TRUE)

#plot loadings
png(filename =  glue::glue("{outfile_ext}_PCA_loadings.POP.png"), width = 8, height = 6, units = "in", res = 300) 
par(mfrow=c(2,1))
  plot(pca_snprelate_load$snploading[1,], type="h", ylab="PC 1")
  plot(pca_snprelate_load$snploading[2,], type="h", ylab="PC 2")
dev.off()


###By SPP

#perform and plot PCA
png(filename =  glue::glue("{outfile_ext}_GRM.SPP.png"), width = 7, height = 4, units = "in", res = 300) 
  PCA_SNA  <- PCA_manual_colors(GRM$grm, 
                                sample_info$IND,
                                sample_info$SPP,
                                1, 2,            #PCs to plot
                                show.point=T, 
                                show.label=F, 
                                manual_colors = spp_colors, 
                                manual_shapes = spp_shapes,
                                show.ellipse=T, 
                                show.line=F, 
                                alpha=0) 
dev.off()

#perform classic PCA - results should the same as the GRM

#first make sure populations are in same order as data frame for plotting/legend order
sample_info$SPP <- factor(sample_info$SPP, levels = unique(sample_info$SPP))
#create data frame containing sample.id, population and EV1 and EV2 (first and second eigenvectors from pca_snprelate)
tab <- data.frame(sample.id  = pca_snprelate$sample.id,
                  region     = sample_info$SPP,
                  EV1        = pca_snprelate$eigenvect[,1],    # the first eigenvector
                  EV2        = pca_snprelate$eigenvect[,2],    # the second eigenvector
                  stringsAsFactors = FALSE)
#selects values where EV1 <= -0.1 and EV2 <= -0.3, but not sure why
tab_select <- dplyr::filter(tab,EV1 <=  -0.1| EV2 <= -0.3)

png(filename =  glue::glue("{outfile_ext}_PCA.SPP.png"), width = 8, height = 4, units = "in", res = 300) 
  ggplot(tab,aes(x=EV1,y=EV2, color = region ))+
    geom_point()+
    geom_text_repel(data = tab_select,aes(x=EV1,y=EV2, color = tab_select$region, label = tab_select$sample.id), show.legend = F,max.overlaps = 30)+
    xlab(glue("PC1 ({round(pc.percent[1],2)}%)"))+
    ylab(glue("PC2 ({round(pc.percent[2],2)}%)")) +
    theme(
      panel.border = element_rect(color = "black", fill = NA, linewidth = 0.75),
      panel.background = element_blank(),
      panel.grid.major = element_line(color = "grey", linewidth = 0.25),
      axis.text.x = element_blank(),
      axis.text.y = element_blank(),
      legend.key = element_blank()) +
    labs(color = "Population")
dev.off()
  
#calculate SNP loadings from PCA
pca_snprelate_load  <- snpgdsPCASNPLoading(pca_snprelate, gds, num.thread=1L, verbose=TRUE)

#plot loadings
png(filename =  glue::glue("{outfile_ext}_PCA_loadings.SPP.png"), width = 8, height = 6, units = "in", res = 300) 
par(mfrow=c(2,1))
  plot(pca_snprelate_load$snploading[1,], type="h", ylab="PC 1")
  plot(pca_snprelate_load$snploading[2,], type="h", ylab="PC 2")
dev.off()
```

DAPC
```{r}
###By POP
#make sure populations are in same order as data frame for plotting/legend order
sample_info$POP <- factor(sample_info$POP, levels = unique(sample_info$POP))

#run DAPC
dapc <- adegenet::dapc(genlight, as.factor(sample_info$POP), n.pca = 10, n.da = 10)

dapc_plot <- function(dapc,file_ext){
    png(file_ext, width = 12, height = 6, units = "in", res = 300)
    ade4::scatter(dapc,  xax  = 1,     yax = 2,
                  scree.da  = T,
                  scree.pca = T, 
                  mstree    = F,   
                  legend    = T,
                  clabel    = F, 
                  cellipse  = 1,
                  cstar     = 0,
                  posi.leg  = "topright"  ,
                  posi.pca  = "topleft",
                  posi.da  = "bottomleft", 
                  cex       = 2, 
                  cex.lab   = 4,
                  lwd       = 3, 
                  lty       = 2,
                  col       = colors,
                  pch       = shapes)
    dev.off()
}


dapc_plot(dapc,glue::glue("{outfile_ext}_DAPC.POP.png"))


###By SPP - DOESN'T WORK
#make sure populations are in same order as data frame for plotting/legend order
sample_info$SPP <- factor(sample_info$SPP, levels = unique(sample_info$SPP))

#run DAPC
spp_dapc <- adegenet::dapc(genlight, as.factor(sample_info$SPP), n.pca = 10, n.da = 10)

spp_dapc_plot <- function(spp_dapc,file_ext){
    png(file_ext, width = 12, height = 6, units = "in", res = 300)
    ade4::scatter(spp_dapc,  xax  = 1,     yax = 2,
                  scree.da  = T,
                  scree.pca = T, 
                  mstree    = F,   
                  legend    = T,
                  clabel    = F, 
                  cellipse  = 1,
                  cstar     = 0,
                  posi.leg  = "topright"  ,
                  posi.pca  = "topleft",
                  posi.da  = "bottomleft", 
                  cex       = 2, 
                  cex.lab   = 4,
                  lwd       = 3, 
                  lty       = 2,
                  col       = spp_colors,
                  pch       = spp_shapes)
    dev.off()
}


spp_dapc_plot(spp_dapc,glue::glue("{outfile_ext}_DAPC.SPP.png"))


###By GROUP - with DI separate to rest of Gibson's
#make sure populations are in same order as data frame for plotting/legend order
sample_info$island <- factor(sample_info$island, levels = unique(sample_info$island))

#run DAPC
island_dapc <- adegenet::dapc(genlight, as.factor(sample_info$island), n.pca = 10, n.da = 10)

island_dapc_plot <- function(island_dapc,file_ext){
    png(file_ext, width = 12, height = 6, units = "in", res = 300)
    ade4::scatter(island_dapc,  xax  = 1,     yax = 2,
                  scree.da  = T,
                  scree.pca = T, 
                  mstree    = F,   
                  legend    = T,
                  clabel    = F, 
                  cellipse  = 1,
                  cstar     = 0,
                  posi.leg  = "topright"  ,
                  posi.pca  = "topleft",
                  posi.da  = "bottomleft", 
                  cex       = 1, 
                  cex.lab   = 4,
                  lwd       = 3, 
                  lty       = 2,
                  col       = island_colors,
                  pch       = island_shapes)
    dev.off()
}


island_dapc_plot(island_dapc,glue::glue("{outfile_ext}_DAPC.island.png"))

```
FST

```{r}
###By POP
#calculate FST between pops
#QQplots are generated when 'out' is specified so you can check assumptions
#seems to be an issue with saving qqplots to outfile_ext, changed directory and saved them elsewhere then manually moved them
PWFST_df <- snpgdsPWFST(gds = gds,
                        pop.id = sample_info$POP,
                        reps = 10000,
                        signf = 0.05,
                        missing.rate = 0.95,
                        #snp.id = SNPs_neutral)
                        )

PWFST_df <- snpgdsPWFST(gds = gds,
                        pop.id = sample_info$POP,
                        reps = 10000,
                        signf = 0.05,
                        missing.rate = 0.95,
                        #snp.id = SNPs_neutral,
                        out = "C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Desktop/")

#convert FST estimates to matrix
PWFST_mat <- snpgdsFSTdf2mat(pwFST_df = PWFST_df) 
PWFST_mat[PWFST_mat < 0] <- 0
colnames(PWFST_mat) <- str_replace_all(colnames(PWFST_mat),"_"," ")
rownames(PWFST_mat) <- str_replace_all(rownames(PWFST_mat),"_"," ")

#save tsv
write.csv(PWFST_df, paste0(outfile_ext,"_pwFST.POP.csv"))


#create heat maps
#set legend breaks manually
legend_breaks <- c(0, 0.01, 0.02, 0.03, 0.04, 0.05)
#create heatmap png
pheatmap::pheatmap(PWFST_mat,cutree_rows = 1
                   ,cutree_cols = 1
                   ,display_numbers = T
                   ,number_format = "%.5f"
                   ,filename = glue::glue("{outfile_ext}_heatmapplot.POP.png")
                   ,width = 12
                   ,height = 8
                   ,fontsize_number = 14
                   #you can add a legend title, but you'd have to set the legend break yourself..
                   ,legend_breaks = c(legend_breaks, max(PWFST_mat))
                   ,legend_labels = c(legend_breaks, "FST\n")
)
#create heatmap pdf
pheatmap::pheatmap(PWFST_mat,cutree_rows = 1
                   ,cutree_cols = 1
                   ,display_numbers = T
                   ,number_format = "%.5f"
                   ,filename = glue("{outfile_ext}_heatmapplot.POP.pdf")
                   ,width = 8
                   ,height = 8
                   #you can add a legend title, but you'd have to set the legend break yourself..
                   ,legend_breaks = c(legend_breaks, max(PWFST_mat))
                   ,legend_labels = c(legend_breaks, "FST\n")
)



###By SPP
#calculate FST between pops
#QQplots are generated when 'out' is specified so you can check assumptions
#seems to be an issue with saving qqplots to outfile_ext, changed directory and saved them elsewhere then manually moved them
PWFST_df <- snpgdsPWFST(gds = gds,
                        pop.id = sample_info$SPP,
                        reps = 10000,
                        signf = 0.05,
                        missing.rate = 0.95,
                        #snp.id = SNPs_neutral,)
                        )

PWFST_df <- snpgdsPWFST(gds = gds,
                        pop.id = sample_info$SPP,
                        reps = 10000,
                        signf = 0.05,
                        missing.rate = 0.95,
                        #snp.id = SNPs_neutral,
                        out = "C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Desktop/qq")

#convert FST estimates to matrix
PWFST_mat <- snpgdsFSTdf2mat(pwFST_df = PWFST_df) 
PWFST_mat[PWFST_mat < 0] <- 0
colnames(PWFST_mat) <- str_replace_all(colnames(PWFST_mat),"_"," ")
rownames(PWFST_mat) <- str_replace_all(rownames(PWFST_mat),"_"," ")

#save tsv
write.csv(PWFST_df, paste0(outfile_ext,"_pwFST.SPP.csv"))

#create heat maps

#create heatmap png
pheatmap::pheatmap(PWFST_mat,cutree_rows = 1
                   ,cutree_cols = 1
                   ,display_numbers = T
                   ,number_format = "%.5f"
                   ,filename = glue::glue("{outfile_ext}_heatmapplot.SPP.png")
                   ,width = 12
                   ,height = 8
                   ,fontsize_number = 14
                   #you can add a legend title, but you'd have to set the legend break yourself..
                   ,legend_breaks = c(legend_breaks, max(PWFST_mat))
                   ,legend_labels = c(legend_breaks, "FST\n")
)
#create heatmap pdf
pheatmap::pheatmap(PWFST_mat,cutree_rows = 1
                   ,cutree_cols = 1
                   ,display_numbers = T
                   ,number_format = "%.5f"
                   ,filename = glue("{outfile_ext}_heatmapplot.SPP.pdf")
                   ,width = 8
                   ,height = 8
                   #you can add a legend title, but you'd have to set the legend break yourself..
                   ,legend_breaks = c(legend_breaks, max(PWFST_mat))
                   ,legend_labels = c(legend_breaks, "FST\n")
)


###By island
#calculate FST between pops
#QQplots are generated when 'out' is specified so you can check assumptions
#seems to be an issue with saving qqplots to outfile_ext, changed directory and saved them elsewhere then manually moved them
PWFST_df <- snpgdsPWFST(gds = gds,
                        pop.id = sample_info$island,
                        reps = 10000,
                        signf = 0.05,
                        missing.rate = 0.95,
                        #snp.id = SNPs_neutral,)
                        )

PWFST_df <- snpgdsPWFST(gds = gds,
                        pop.id = sample_info$SPP,
                        reps = 10000,
                        signf = 0.05,
                        missing.rate = 0.95,
                        #snp.id = SNPs_neutral,
                        out = "C:/Users/footeim/OneDrive - Victoria University of Wellington - STAFF/Desktop/qq")

#convert FST estimates to matrix
PWFST_mat <- snpgdsFSTdf2mat(pwFST_df = PWFST_df) 
PWFST_mat[PWFST_mat < 0] <- 0
colnames(PWFST_mat) <- str_replace_all(colnames(PWFST_mat),"_"," ")
rownames(PWFST_mat) <- str_replace_all(rownames(PWFST_mat),"_"," ")

#save tsv
write.csv(PWFST_df, paste0(outfile_ext,"_pwFST.island.csv"))

#create heat maps
#set legend breaks manually

#create heatmap png
pheatmap::pheatmap(PWFST_mat,cutree_rows = 1
                   ,cutree_cols = 1
                   ,display_numbers = T
                   ,number_format = "%.5f"
                   ,filename = glue::glue("{outfile_ext}_heatmapplot.island.png")
                   ,width = 12
                   ,height = 8
                   ,fontsize_number = 14
                   #you can add a legend title, but you'd have to set the legend break yourself..
                   ,legend_breaks = c(legend_breaks, max(PWFST_mat))
                   ,legend_labels = c(legend_breaks, "FST\n")
)
#create heatmap pdf
pheatmap::pheatmap(PWFST_mat,cutree_rows = 1
                   ,cutree_cols = 1
                   ,display_numbers = T
                   ,number_format = "%.5f"
                   ,filename = glue("{outfile_ext}_heatmapplot.island.pdf")
                   ,width = 8
                   ,height = 8
                   #you can add a legend title, but you'd have to set the legend break yourself..
                   ,legend_breaks = c(legend_breaks, max(PWFST_mat))
                   ,legend_labels = c(legend_breaks, "FST\n")
)

```


Plot pairwise relatedness/divergence?

```{r}
GRM_matrix <- GRM$grm
heatmap(GRM_matrix,
        symm = TRUE,        # Ensure the heatmap is symmetric
        col = colorRampPalette(c("blue", "white", "red"))(100),  # Color palette
        main = "Pairwise Genetic Relatedness",  # Title of the plot
        xlab = "Sample ID",  # Label for x-axis
        ylab = "Sample ID",  # Label for y-axis
        margins = c(10, 10)  # Adjust margins for better readability
)

plot(GRM$grm[44:78,44:78])


gl.dist <- gl.dist.ind(genlight, method="Euclidean", scale = TRUE, output = "matrix", plot.out = TRUE)
upper_triangle <- upper.tri(gl.dist, diag = FALSE)
gl.dist[upper_triangle] <- NA
upper_half_matrix <- matrix(gl.dist, nrow = nrow(gl.dist), ncol = ncol(gl.dist))

png(glue::glue("{outfile_ext}_genetic_distance.png"))
  plot(upper_half_matrix, xlim=c(0.35, 0.45), ylim=c(0.35,0.45))
dev.off()


GRM_matrix <- as.matrix(GRM$grm)
upper_triangle_GRM <- upper.tri(GRM_matrix, diag = FALSE)
GRM_matrix[upper_triangle] <- NA
upper_half_matrix_GRM <- matrix(GRM_matrix, nrow = nrow(GRM_matrix), ncol = ncol(GRM_matrix))
plot(upper_half_matrix_GRM)

```





